### PR TITLE
OIDC code exchange add private key JWT token exchange

### PR DESCRIPTION
**Pull request by sheidkamp**
# Description
API changes to support PK JWT support.

On the public interface, we are currently only supporting Client Secret token exchange, and the OIDC configuration looks like:
```
oidcAuthorizationCode:
  appUrl: https://glootest.com
  callbackPath: /callback
  clientId: <insert-your-app-client-id-here>
  clientSecretRef:
    name: auth0-client-secret
    namespace: gloo-system
  disableClientSecret: false
  issuerUrl: https://solo-io.us.auth0.com/
  scopes:
  - email
```

We will deprecate and continue to support this format, but the preferred format will be to move the configuration related to the client authentication to its own block:
```
oidcAuthorizationCode:
  appUrl: https://glootest.com:32500
  callbackPath: /callback
  clientId: qlehSNyVlbunoGlTcOKygdNvKuf72j0k
  clientAuthentication:
    clientSecret:
      clientSecretRef:
      name: auth0-client-secret
    disableClientSecret: false
  namespace: gloo-system
  issuerUrl: https://dev-2e4epk7rzc1uoat8.us.auth0.com/
  scopes:
  - email
 ```
The private key configuration will then look like:
```
oidcAuthorizationCode:
  appUrl: https://glootest.com:32500
  callbackPath: /callback
  clientId: qlehSNyVlbunoGlTcOKygdNvKuf72j0k
  clientAuthentication:
    privateKeyJwt:
      signingKeyRef:
        name: auth0-client-secret-jwt
        namespace: gloo-system
      validFor: 1s
  namespace: gloo-system
  issuerUrl: https://dev-2e4epk7rzc1uoat8.us.auth0.com/
  scopes:
  - email
 ```

In the internal API, we will add a top-level, optional `pk_jwt_exchange_config` field to the `OidcAuthorizationCodeConfig` message to pass the PK JWT configuration if it exists, and we will rely on our translation and validation code to ensure that   at most one of `pk_jwt_exchange_config` and `client_secret` is defined. (`client_sceret` may be null if `disableClientSecret` is set to true)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 1]
```diff
@@ -704,6 +706,39 @@ message OidcAuthorizationCode {
...
+            google.protobuf.BoolValue disable_client_secret = 2;
+        }
+
+         // Client Secret Exchange requires a signing key for the JWT and an duration for the JWT to be valid.
+        message PrivateKeyJwt{
+            // Signing key for the JWT used to exchange the access code for the tokens
+            core.solo.io.ResourceRef signing_key_ref = 1 [(extproto.sensitive) = true];
+            // Amount of time for which the JWT is valid. No maximmum is enforced, but different IDPs may impose limits on how far in 
+            // the future the expiration time is allowed to be. If omitted, default is 5s.
+            google.protobuf.Duration valid_for = 2;
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:729]: I think you mentioned this during the demo today, but I just wanted to confirm: were you able to find any docs about recommended durations, and the implications of jwts that are minted for too long?

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 2]
```diff
@@ -1510,6 +1545,13 @@ message ExtAuthConfig {
 
         // Configuration related to the user session.
         UserSessionConfig user_session = 20;
+
+        message PkJwtExchangeConfig{
+            string signing_key = 1 [(extproto.sensitive) = true];
+            google.protobuf.Duration valid_for = 2;
+        }
+
+        PkJwtExchangeConfig pk_jwt_exchange_config = 22;
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1554]: nit: I don't see a field for 21, am I missing something?

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 3]
```diff
@@ -1510,6 +1545,13 @@ message ExtAuthConfig {
 
         // Configuration related to the user session.
         UserSessionConfig user_session = 20;
+
+        message PkJwtExchangeConfig{
+            string signing_key = 1 [(extproto.sensitive) = true];
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1550]: nit: Let's add comments here, even if they mirror the comments on the user-facing API

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 4]
```diff
@@ -569,7 +569,8 @@ message OidcAuthorizationCode {
 
     // your client secret as registered with the issuer.
     // This is required unless `disable_client_secret` is true
-    core.solo.io.ResourceRef client_secret_ref = 2;
+    // This field has been deprecated and can be set in the client_secret option of code_exchange_type
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:572]: nit: I think this comment may be out of date. I don't see a code_exchange_type, did you mean client_authentication?

File: changelog/v1.16.0-beta12/oidc-code-exchange-refactor.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+changelog:
+- type: NON_USER_FACING
+  issueLink: https://github.com/solo-io/gloo/issues/7744
+  description: Refactor the internal extauth service interface in preparation for adding PK JWT exchange support
```
Comment (sam-heilbron) [inline changelog/v1.16.0-beta12/oidc-code-exchange-refactor.yaml:4]: Is this accurate?

Comment (sheidkamp) [reply changelog/v1.16.0-beta12/oidc-code-exchange-refactor.yaml:4]: nope, updated

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 4 above)
Comment (sheidkamp) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto]: correct

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 6]
```diff
@@ -704,6 +706,39 @@ message OidcAuthorizationCode {
...
         repeated ClaimToHeader claims_to_headers = 1;
     }
+
+    // Configuration specific to the code exchange type used to exchange the access code for the access and id tokens.
+    message ClientAuthentication {
+        // Client Secret Exchange requires a client secret (unless it is disabled)
+        message ClientSecret {
+            // your client secret as registered with the issuer.
+            // This is required unless `disable_client_secret` is true
+            core.solo.io.ResourceRef client_secret_ref = 1 [(extproto.sensitive) = true];
```
[the remark this answers is not in the document]
Comment (sheidkamp) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto]: you are correct, this is not needed here.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 1 above)
Comment (sheidkamp) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:729]: I wasn't able find anything other than "keep is as short as you can", and should account for latency and a small amount of clock skew. 5s seems reasonable, and I am happy to adjust it.

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 2 above)
Comment (sheidkamp) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto]: ah, side effect from when that was nested

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto (hunk 3 above)
Comment (sheidkamp) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1552]: Added

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 7]
```diff
@@ -683,7 +684,8 @@ message OidcAuthorizationCode {
     // If true, do not check for or use the client secret.
     // Generally the client secret is required and AuthConfigs will be rejected if it isn't set.
     // However certain implementations of the PKCE flow do not use a client secret (including Okta) so this setting allows configuring Oidc without a client secret.
-    google.protobuf.BoolValue disable_client_secret = 21;
+    // This field has been deprecated and can be set in the client_secret option of code_exchange_type
```
Comment (sam-heilbron) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:687]: nit: same as the above comment that you updated, I think we need to update this to not reference the code_exchange_type

Comment (sheidkamp) [reply projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto]: updated

**Review by sam-heilbron — APPROVED** [review]

**Review by inFocus7 — DISMISSED** [review]

File: projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto [hunk 8]
```diff
@@ -1510,6 +1545,19 @@ message ExtAuthConfig {
 
         // Configuration related to the user session.
         UserSessionConfig user_session = 20;
+
+        // Fields for private key JWT exchange.
+        message PkJwtExchangeConfig{
```
Comment (sheidkamp) [inline projects/gloo/api/v1/enterprise/options/extauth/v1/extauth.proto:1550]: Adding a comment so it doesn't merge... I want to rename this to get away from the "Exchange" terminology.

**Review by sam-heilbron — APPROVED** [review]

**Review by inFocus7 — APPROVED** [review]

**Pull request merged**